src/parcsr_ls/par_lr_restr_device.c

```diff
@@ -0,0 +1,379 @@
+   HYPRE_BigInt       *col_map_offd_R;
+   HYPRE_Int           num_cols_offd_R;
+   HYPRE_Int           my_id, num_procs;
+   HYPRE_BigInt        total_global_cpts;
+   HYPRE_Int           nnz_diag, nnz_offd;
+   HYPRE_BigInt       *send_buf_i;
+   HYPRE_Int           i;
+   HYPRE_Int          *CF_marker;
+
+   /* local size */
+   HYPRE_Int n_fine = hypre_CSRMatrixNumRows(A_diag);
+   HYPRE_BigInt col_start = hypre_ParCSRMatrixFirstRowIndex(A);
+
+   /* MPI size and rank*/
+   hypre_MPI_Comm_size(comm, &num_procs);
+   hypre_MPI_Comm_rank(comm, &my_id);
+
+   /* copy CF_marker to the device if necessary */
+   hypre_MemoryLocation cf_memory_location;
+   hypre_GetPointerLocation(CF_marker_host, &cf_memory_location);
+   if (cf_memory_location == hypre_GetActualMemLocation(HYPRE_MEMORY_HOST))
+   {
+      CF_marker = hypre_TAlloc(HYPRE_Int, n_fine, HYPRE_MEMORY_DEVICE);
+      hypre_TMemcpy(CF_marker, CF_marker_host, HYPRE_Int, n_fine, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_HOST);
+   }
+   else
+   {
+      CF_marker = CF_marker_host;
+   }
+
+   /* global number of C points and my start position */
+   if (my_id == (num_procs -1))
+   {
+      total_global_cpts = num_cpts_global[1];
+   }
+   hypre_MPI_Bcast(&total_global_cpts, 1, HYPRE_MPI_BIG_INT, num_procs-1, comm);
+
+   /* get AFF and ACF */
+   hypre_ParCSRMatrix *AFF, *ACF, *Dinv, *N, *X, *X2, *Z, *Z2;
+   if (strong_thresholdR > 0)
+   {
+      hypre_ParCSRMatrix *S;
+      hypre_BoomerAMGCreateSabs(A,
+                                strong_thresholdR,
+                                0.9,
+                                num_functions,
+                                dof_func,
+                                &S);
+      hypre_ParCSRMatrixGenerateFFCFDevice(A, CF_marker, num_cpts_global, S, &ACF, &AFF);
+      hypre_ParCSRMatrixDestroy(S);
+   }
+   else
+   {
+      hypre_ParCSRMatrixGenerateFFCFDevice(A, CF_marker, num_cpts_global, NULL, &ACF, &AFF);
+   }
+
+   HYPRE_Int        n_fpts = hypre_CSRMatrixNumRows(hypre_ParCSRMatrixDiag(AFF));
+   HYPRE_Int        n_cpts = n_fine - n_fpts;
+   hypre_assert(n_cpts == hypre_CSRMatrixNumRows(hypre_ParCSRMatrixDiag(ACF)));
+
+   /* maps from F-pts and C-pts to all points */
+   HYPRE_Int       *Fmap = hypre_TAlloc(HYPRE_Int, n_fpts, HYPRE_MEMORY_DEVICE);
+   HYPRE_Int       *Cmap = hypre_TAlloc(HYPRE_Int, n_cpts, HYPRE_MEMORY_DEVICE);
+   HYPRE_THRUST_CALL( copy_if,
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(n_fine),
+                      CF_marker,
+                      Fmap,
+                      is_negative<HYPRE_Int>());
+   HYPRE_THRUST_CALL( copy_if,
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(n_fine),
+                      CF_marker,
+                      Cmap,
+                      is_positive<HYPRE_Int>());
+
+   /* setup Dinv = 1/(diagonal of AFF) */
+   Dinv = hypre_ParCSRMatrixCreate(comm,
```

Comment:
@waynemitchell Using matrix-matrix multiplication for diagonal scaling seems to be a too big hammer for me (especially on GPUs). We have done similar scalings for interpolation, see `hypreCUDAKernel_compute_aff_afc`. Please let me know if I can help with anything.

Reply:
I had thought about this a while ago. I opted for mat-mat multiplication just for ease of implementation on the first try, especially since we multiply both on the left and the right by Dinv at different places in the algorithm. Then I just forgot about ever going back to write something more optimized. I'll take a look at this.

Reply:
Thanks @waynemitchell. I just wanted to point it out. We can leave the optimization for future. Your call.

Reply:
@liruipeng I think I'll leave this for a future PR, just to get this one finished up.

Reply:
@waynemitchell Sounds good to me. I actually found some "diagonal scaling" routine (from both the left and right). I can add a `ParcsrDiagScale` later. Thanks!